proposals/0239-operator-designated-types.md

```diff
@@ -0,0 +1,137 @@
+
+*During the review process, add the following fields as needed:*
+
+* Implementation: This is currently in-tree, disabled by default, and enabled by staging options (`-enable-operator-designated-types`, `-solver-enable-operator-designated-types`).
+* Decision Notes: [Rationale](https://forums.swift.org/), [Additional Commentary](https://forums.swift.org/)
+* Bugs: [SR-NNNN](https://bugs.swift.org/browse/SR-NNNN), [SR-MMMM](https://bugs.swift.org/browse/SR-MMMM)
+* Previous Revision: [1](https://github.com/apple/swift-evolution/blob/...commit-ID.../proposals/NNNN-filename.md)
+* Previous Proposal: [SE-XXXX](XXXX-filename.md)
+
+## Introduction
+
+Swift's expression type checker is known to have cases where it is extremely slow in type checking expressions.
+
+Some of these slow typechecking cases involve expressions that make use of operators (whereas others involve collections or chained single-expression closures).
+
+This proposal addresses speeding up type checking for cases that are slow primarily due to the number of combinations of overloads of operators that are examined in the course of typechecking.
+
+Swift-evolution thread: [Pitch: Making expression type checking of operator expressions fast](https://forums.swift.org/t/pitch-making-expression-type-checking-of-operator-expressions-fast/18037)
+
+## Motivation
+
+Users sometimes write reasonable looking expressions that result in either the compiler taking a very long time to typecheck the expression or results in the diagnostic `the compiler is unable to type-check this expression in reasonable time` without any further hints as to how to fix the problem.
+
+There are various reasons users hit this error:
+
+- Expressions involving many operators and (usually) several literals, that should successfully typecheck.
+- Expressions that should not successfully typecheck (but fail to emit a good diagnostic prior to the type checker giving up).
+- Expressions involving collection literals sometimes fail even if they have only a handful of elements (e.g. < 20).
+- Chains of single-expression closures, which in addition to exacerbating the issues with operators and literals, also create new problems due to a lack of explicit types for their parameters.
+
+This proposal aims to address only the first of these issues.
+
+## Proposed solution
+
+We introduce syntax to define *designated types* in operator declarations. These types will be used as the first place to look up operator overloads for a given operator. If the expression successfully typechecks with an overload found via those types, we stop looking at the remaining overloads for the operator.
+
+The intent is to use this within `Policy.swift` in the standard library in order to provide a hint to the type checker as to which overload should be preferred for typechecking an operator. You can see the current usage of this syntax [here](https://github.com/apple/swift/blob/master/stdlib/public/core/Policy.swift#L398).
```

Comment:
Use permalink? e.g. [here](https://github.com/apple/swift/blob/191a71e10dab37aa12b53f04b9496642dd3604b1/stdlib/public/core/Policy.swift#L382-L485).

Reply:
@benrimmington Fixed in https://github.com/apple/swift-evolution/pull/966/commits/c051e12766d568dab70002374fdfccce0d8d0467, thanks!